pkg/alertmanager/distributor_test.go

```diff
@@ -0,0 +1,297 @@
+	"github.com/weaveworks/common/user"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health/grpc_health_v1"
+)
+
+func TestDistributor_DistributeRequest(t *testing.T) {
+	cases := []struct {
+		name                string
+		numAM, numHappyAM   int
+		replicationFactor   int
+		isRead              bool
+		expStatusCode       int
+		expectedTotalCalls  int
+		headersNotPreserved bool
+	}{
+		{
+			name:               "Simple AM request, all AM healthy",
+			numAM:              4,
+			numHappyAM:         4,
+			replicationFactor:  3,
+			expStatusCode:      http.StatusOK,
+			expectedTotalCalls: 3,
+		}, {
+			name:                "Less than quorum AM available",
+			numAM:               1,
+			numHappyAM:          1,
+			replicationFactor:   3,
+			expStatusCode:       http.StatusInternalServerError,
+			expectedTotalCalls:  0,
+			headersNotPreserved: true, // There is nothing to preserve since it does not hit any AM.
+		}, {
+			name:               "Less than quorum AM succeed",
+			numAM:              5,
+			numHappyAM:         3, // Though we have 3 happy, it will hit >1 unhappy AM.
+			replicationFactor:  3,
+			expStatusCode:      http.StatusInternalServerError,
+			expectedTotalCalls: 3,
+		}, {
+			name:               "Read is sent to only 1 AM",
+			numAM:              5,
+			numHappyAM:         5,
+			replicationFactor:  3,
+			isRead:             true,
+			expStatusCode:      http.StatusOK,
+			expectedTotalCalls: 1,
+		},
+	}
+
+	route := "/alertmanager/api/v1/alerts"
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			d, ams, cleanup := prepare(t, c.numAM, c.numHappyAM, c.replicationFactor)
+			t.Cleanup(cleanup)
+
+			ctx := user.InjectOrgID(context.Background(), "1")
+
+			url := "http://127.0.0.1:9999" + route
+			req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader([]byte{1, 2, 3, 4}))
+			require.NoError(t, err)
+			if c.isRead {
+				req.Method = http.MethodGet
+			}
+			req.RequestURI = url
+
+			w := httptest.NewRecorder()
+			d.DistributeRequest(w, req)
+			resp := w.Result()
+
+			require.Equal(t, c.expStatusCode, resp.StatusCode)
+
+			if !c.headersNotPreserved {
+				// Making sure the headers are not altered.
+				contentType := []string{"it-is-ok"}
+				contentTypeOptions := []string{"ok-option-1", "ok-option-2"}
+				if resp.StatusCode != http.StatusOK {
+					contentType = []string{"it-is-not-ok"}
+					contentTypeOptions = []string{"not-ok-option-1", "not-ok-option-2"}
+				}
+				require.Equal(t, contentType, resp.Header.Values("Content-Type"))
+				require.Equal(t, contentTypeOptions, resp.Header.Values("X-Content-Type-Options"))
+			}
+
+			// Since the response is sent as soon as the quorum is reached, when we
+			// reach this point the 3rd AM may not have received the request yet.
+			// To avoid flaky test we retry until we hit the desired state within a reasonable timeout.
+			test.Poll(t, time.Second, true, func() interface{} {
```

Comment:
This and the next suggestion below improves the error reporting on failure (you will see the actual value instead of just `false`):
```suggestion
			test.Poll(t, time.Second, c.expectedTotalCalls, func() interface{} {
```